### PR TITLE
MAINT: pre-process/transform the data before training

### PR DESCRIPTION
**Pull request by stsievert**
**What does this PR implement?**
It performs the data transformations before any training begins. If a PyTorch dataset is passed to `.fit`, these transformations will be completed once every epoch (which might be slow). If the user wants to avoid that, they should use this code:

``` python
from adadamp import DaskBaseDamper
from torch.utils.data import transforms
from torchvision import datasets, transforms

net = DaskClassifier(module=Model, ...)

transform = transforms.Compose(
    [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
)
ds = datasets.MNIST("./data", train=True, download=True, transform=transform)
ds2 = net.preprocess(df)
for epoch in range(10):
    net.partial_fit(ds2)
```

### CODE REVIEW

**Comment by stsievert** [comment]
This is a performance improvement: the transformations took 70% of time spent in `gradient`. I've looked at `decreasing-lr/exp--final.csv`, and here are the predicted timings:

|                | Avg. time for<br>`partial_fit`  | Equation |
| --------:     | :--- |:--- |
| Before this PR | 207.7ms | `t = partial_fit__time / n_updates` |
| After this PR  | ~62ms  | `t * (1 - 0.7)` |

<details>
<summary>Source equations</summary>

``` python
import pandas as pd
df = pd.read_csv("stats/decreasing-lr/exp--final.csv")
d = df[["partial_fit__time", "n_updates"]].copy()
running_avg = d["partial_fit__time"].cumsum() / df["n_updates"]
print(running_avg.iloc[-1])  # 0.207651...
```

I found that a call to gradient spent 70% of the time doing transformations by looking at `dask-report.html`, which says for the entire training, 810.76s were spent performing transformations and 1148.83s were spent in `gradient`, and `810 / 1148` is approximately `0.7055`.

</details>

I have a hard time believing this performance; 60ms is pretty fast. @JoeHolt can you spin up an EC2 instance and run a couple epochs (and save the CSV) to verify that the average time for a call to `gradient` is 60ms?

**Comment by stsievert** [comment]
I got the time for a round-trip model update down to 147ms. Here's the breakdown:

| Function | Time (ms) |
|:---|:------|
| `update_model` | 11 |
| `gradient`, deepcopy | 58 |
| `gradient` calculation | 78 |
| **Total** | 147 |

I think a comparison with the naive implementation should be generated. I think it'd be sufficient to take PyTorch's MNIST example and including this model/dataset. This would only be a timing comparison, so loss/optimizer/hyperparameters can be ignored.

I tried to compare with https://github.com/soumith/convnet-benchmarks/, but there are important size differences (image size, batch size, network size). It's in the same ballpark, but I can't tell if that's reasonable.
